app/src/lib/highlighter/worker.ts

```diff
@@ -0,0 +1,85 @@
+import { ITokens, IHighlightRequest } from './types'
+
+const highlightWorkers = new Array<Worker>()
+const maxIdlingWorkers = 2
+const workerMaxRunDuration = 5 * 1000
+
+/**
+ * Request an automatic detection of the language and highlight
+ * the contents provided.
+ *
+ * @param contents  The actual contents which is to be used for
+ *                  highlighting.
+ * @param extension The file extension of the path in question as returned
+ *                  by node's extname() function (i.e. with a leading dot).
+ * @param tabSize   The width of a tab character. Defaults to 4. Used by the
+ *                  stream to count columns. See CodeMirror's StringStream
+ *                  class for more details.
+ * @param lines     An optional filter of lines which needs to be tokenized.
+ *
+ *                  If undefined or empty all lines will be tokenized
+ *                  and returned. By passing an explicit set of lines we can
+ *                  both minimize the size of the response object (which needs
+ *                  to be serialized over the IPC boundary) and, for stateless
+ *                  modes we can significantly speed up the highlight process.
+ */
+export function highlight(
+  contents: string,
+  extension: string,
+  tabSize: number,
+  lines: Array<number>
+): Promise<ITokens> {
+  // Bail early if there's no content to highlight or if we don't
+  // need any lines from this file.
+  if (!contents.length || !lines.length) {
+    return Promise.resolve({})
+  }
+
+  // Get an existing, idle worker or create a new one if none
+  // exists.
+  const worker =
+    highlightWorkers.shift() ||
+    new Worker(`file:///${__dirname}/highlighter.js`)
```

Comment:
Should we be using `encodePathAsUrl` here, ala https://github.com/desktop/desktop/pull/3109?

Reply:
Ah, yes. That happened after I wrote this 👍